lxd/daemon.go

```diff
@@ -39,15 +39,20 @@ import (
 
 // A Daemon can respond to requests from a shared client.
 type Daemon struct {
-	clientCerts         []x509.Certificate
-	os                  *sys.OS
-	db                  *sql.DB
-	readyChan           chan bool
-	pruneChan           chan bool
-	shutdownChan        chan bool
-	resetAutoUpdateChan chan bool
-
-	tasks     task.Group
+	clientCerts  []x509.Certificate
+	os           *sys.OS
+	db           *sql.DB
+	readyChan    chan bool
+	shutdownChan chan bool
+
+	// Tasks registry for long-running background tasks.
+	tasks task.Group
+
+	// Indexes of tasks that need to be reset when their execution interval
+	// changes.
+	taskPruneImages int
+	taskAutoUpdate  int
```

Comment:
This getting an int index back and keeping it around feels a bit weird. I'd have expected to get a pointer to some kind of struct back when adding a task to the scheduler. This struct would then hold the Reset, Cancel, ... functions rather than require you to pass the task index to separate functions.

Reply:
It's been a while so I don't exactly recall my own rational. But most probably it's because the only API you'd have on the task object would be Reset, since for now there's no use case for Cancel (we want to cancel all tasks at once when we shutdown).

As a compromise what about declaring "type TaskHandle int" and than making taskPruneImages a TaskHandle?

Or if you still feel strong about exposing a full-fledged object, I can rework it that way.

Reply:
Yeah, feels like we should just make make task public (Task) and have it export a public Reset() function. So rather than have to call the group's Reset function and pass it the index of the task in the ground, you'd just call Reset on the Task directly.

That'd hopefully let us remove any of that index based code from the task package as looking at it closer, I'm fairly sure it's racy right now (if you call Add twice in parallel, you may end up with collisions on the index).

Reply:
I changed the code to expose Task and added Task.Reset in b9fb82bf48befecf40e52e3a862820737b2c4e3c.

Note however that task.Group.Add is still thread-unsafe, and the use of indexes vs *Task pointers is unrelated, since the append function is not thread-safe in Go. There are plenty of cases where we have thread-unsafe code just because we don't use that code concurrently. Since there's no concurrent use of task.Group.Add yet, I'd leave it like it is (the alternative is to add mutexes).

Reply:
Yeah, that's fine.